File: medium/3310_remove_methods_from_project/solution.py
from collections import deque
from typing import List


class Solution:
    def remainingMethods(
        self, n: int, k: int, invocations: List[List[int]]
    ) -> List[int]:
        calledMap = [[] for _ in range(n)]

        for a, b in invocations:
            calledMap[a].append(b)

        suspicious = [False] * n
        suspicious[k] = True

        dq = deque([k])
        while dq:
            a = dq.popleft()

            for b in calledMap[a]:
                if not suspicious[b]:
                    suspicious[b] = True
                    dq.append(b)

        for a, b in invocations:
            if not suspicious[a] and suspicious[b]:
                return list(range(n))

        return [a for a in range(n) if not suspicious[a]]

File: medium/3310_remove_methods_from_project/solution.cpp
#include <queue>
#include <vector>

class Solution {
public:
  std::vector<int>
  remainingMethods(int n, int k, std::vector<std::vector<int>> &invocations) {
    std::vector<std::vector<int>> calledMap(n);

    for (const auto &invocation : invocations)
      calledMap[invocation[0]].push_back(invocation[1]);

    std::vector<bool> suspicious(n, false);
    suspicious[k] = true;

    std::deque<int> dq = {k};
    while (!dq.empty()) {
      int a = dq.front();
      dq.pop_front();

      for (const int &b : calledMap[a]) {
        if (!suspicious[b]) {
          suspicious[b] = true;
          dq.push_back(b);
        }
      }
    }

    std::vector<int> methods;
    methods.reserve(n);

    for (const auto &invocation : invocations) {
      if (!suspicious[invocation[0]] && suspicious[invocation[1]]) {
        for (int a = 0; a < n; ++a)
          methods.push_back(a);

        break;
      }
    }

    if (methods.empty()) {
      for (int a = 0; a < n; ++a)
        if (!suspicious[a])
          methods.push_back(a);
    }

    return methods;
  }
};
